### ISSUE TITLE
[RFE] Loop through the iterations and send object delete requests with rate limit (instead of deleting namespaces)

### ISSUE THREAD

**Issue by venkataanil** [cloud-bulldozer/kube-burner#452]
**Is your feature request related to a problem? Please describe.**

Currently we rate limit (qps & burst) objects  creation while running the workload. After finishing the test we delete the namespace rather than each object (example deployment).  
Resource usage metrics during creation phase grows lineraly however we see very high resource usage (spikes) during deletion as we try delete all objetcs at once without rate limits. We need to delete resources in the same way that they are created. In this approach we  loop through iterations and delete objects rahter than deleting the namespaces and then analyse how the resource consumtion happening in this method. 

**Describe the solution you'd like**
Loop through the iterations and send object delete requests with rate limit similar to what we are doing for creation.

**Describe alternatives you've considered**

**Additional context**

**Comment by smalleni** [comment]
@venkataanil If we don't set anything, client-go's default's may be getting used.

**Comment by smalleni** [comment]
I would imagine we use client-go's defaults https://github.com/kubernetes/client-go/blob/master/rest/config.go#L44

**Comment by smalleni** [comment]
But this approach would drastically increase resource cleanup time - I'm curious what this does to our existing tests in CPT - it perhaps increases cost/cluster uptime?

**Comment by rsevilla87** [comment]
Just to give you some background info, there're three situations when a namespace cleanup can happen in a `create` job type:
- When the flag cleanup is true and there's stale namespaces from a previous run, this stage is not indexed
https://github.com/cloud-bulldozer/kube-burner/blob/50e045e083c689f5173ed634f01a24222c59f0eb/pkg/burner/job.go#L125
- When garbage collection is enabled, this stage happens once the benchmark is finished and it's not indexed either
https://github.com/cloud-bulldozer/kube-burner/blob/50e045e083c689f5173ed634f01a24222c59f0eb/pkg/burner/job.go#L196-L199
-  When workload churning is enabled, a subset of namespaces configured by `ChurnPercent` will be deleted and recreated
https://github.com/cloud-bulldozer/kube-burner/blob/50e045e083c689f5173ed634f01a24222c59f0eb/pkg/burner/create.go#L318-L319


The function CleanupNamespaces uses the regular client object used for object creation, hence the QPS/Burst rates would be similar.


As @venkataanil suggested, cleaning up namespaces (with all their objects) with the configured QPS/Burst rate can be overkill and it leads to quite high resource usage spikes on several components such as kubelet, CRI-O or the apiserver .

I don't like this behavior either, but we must take a group decision on how to perform this cleanup since a change on this can affect the results in workloads w/ churning enabled.

**Comment by jtaleric** [comment]
>I'm not a big fan of this behavior either, but we must take a group decision on how to perform this cleanup since a change on this can affect the results in workloads w/ churning enabled.

I wouldn't be in favor of changing the current cleanup, but maybe adding a flag to change the behavior of cleanup. IMHO this is a super low priority RFE... and I am not sure the work to do this is worth the effort... What does everyone else think?

**Comment by dry923** [comment]
I'm in the mindset of since we know why it isn't really an issue. It should still be consistent run to run which is the important metric. 

This would likely be more important if we were doing churn on node density tests where there could be far more pods in a single namespace than in cluster density. So a flag might be a good long term option but I would agree w/ @jtaleric that it would be pretty low prio atm.

**Comment by vishnuchalla** [comment]
@venkataanil Is this request still relevant? What are we trying to measure/quantify during the cleanup phase and how is this change going to help? I believe its going to be helpful only in churn cases. We should only consider this if its really going to add value to our tests as we already have a good amount of flags in our configuration. Feel free to let me know your thoughts on this.

**Comment by mukrishn** [comment]
I am agreeing with this RFE :+1:, we need a seamless/more-controlled way of cleaning up resource especially while churn, we run most of the hypershift jobs with churn and they consume a lot of resource during deletion. Also that affects the report calculation when max/mean, there are numerous spike in between and most of it caused by deletion. 

Another workaround could be using a `Delete` JobType but it's an interim hack with custom job template, I know you are using a custom template for OVN-IC tests. In baremetal tests we used define our own delete logic in between job for ICNI benchmarks by templat'ing the specifics(% or n) of load to be removed. We can try a dynamic template like [this](https://github.com/redhat-performance/web-burner/blob/f0c1d9c736ad31a648d690a15142c620ba8a0c8f/workload/cfg_delete_icni2_node_density2.yml) for your workload and see it if it helps. 

But I don't want this RFE to be lower prioritized as I see lot of benefits with this and would be helpful in a CNCF pov.

**Comment by vishnuchalla** [comment]
@venkataanil @mukrishn Thanks for a detailed explanation. Looks like we definitely need this implementation for churn. For rest of the cases we can run and create job with `GC` disabled and a separate delete job next to it.

So shall we go ahead with implementing this for churn cases? cc: @rsevilla87 @jtaleric @dry923 @smalleni

**Comment by vishnuchalla** [comment]
Opened up a PR to handle deletes for namespaced resources with same rate in churn cases: https://github.com/cloud-bulldozer/kube-burner/pull/492. Feel free to share thoughts, if we want to make it a default behaviour for churn or needs to be regulated with a flag. Thank you.
